src/core/core.controller.js

```diff
@@ -852,8 +852,7 @@ class Chart {
   getDatasetMeta(datasetIndex) {
     const dataset = this.data.datasets[datasetIndex];
     const metasets = this._metasets;
-    let meta = metasets.filter(x => x && x._dataset === dataset).pop();
-
+    let meta = metasets[datasetIndex];
```

Comment:
I think this will not work correctly, when a dataset is inserted or deleted etc.

Maybe similar performance boost could be gained by creating a cache that can be used on subsequent calls of update cycle.

Or maybe simply just verify metaset has the correct datasetindex and only do the lookup if it does not.

Reply:
> I think this will not work correctly, when a dataset is inserted or deleted etc.
> 

When datasets are inserted or deleted, update() has to be called anyways. So my current plan as I mentioned before is to recreate the meta sets ONCE per update call.

Then, when chartjs iterates over the datasets, the index lookup can be used as implemented right now.

Reply:
I did it the hard way and debugged in the browser. I got all the all unit tests to work, the image based tests do not work locally, as explained before.